xds/src/main/java/io/grpc/xds/client/XdsClientImpl.java

```diff
@@ -749,4 +914,116 @@ private void notifyWatcher(ResourceWatcher<T> watcher, T update) {
     }
   }
 
+  private class ResponseHandler implements XdsResponseHandler {
+    final ServerInfo serverInfo;
+
+    ResponseHandler(ServerInfo serverInfo) {
+      this.serverInfo = serverInfo;
+    }
+
+    @Override
+    public void handleResourceResponse(
+        XdsResourceType<?> xdsResourceType, ServerInfo serverInfo, String versionInfo,
+        List<Any> resources, String nonce, ProcessingTracker processingTracker) {
+      checkNotNull(xdsResourceType, "xdsResourceType");
+      syncContext.throwIfNotInThisSynchronizationContext();
+      Set<String> toParseResourceNames =
+          xdsResourceType.shouldRetrieveResourceKeysForArgs()
+          ? getResourceKeys(xdsResourceType)
+          : null;
+      XdsResourceType.Args args = new XdsResourceType.Args(serverInfo, versionInfo, nonce,
+          bootstrapInfo, securityConfig, toParseResourceNames);
+      handleResourceUpdate(args, resources, xdsResourceType, processingTracker);
+    }
+
+    @Override
+    public void handleStreamClosed(Status status, boolean shouldTryFallback) {
+      syncContext.throwIfNotInThisSynchronizationContext();
+
+      ControlPlaneClient cpcClosed = serverCpClientMap.get(serverInfo);
+      if (cpcClosed == null) {
+        return;
+      }
+
+      cleanUpResourceTimers(cpcClosed);
+
+      if (status.isOk()) {
+        return; // Not considered an error
+      }
+
+      Collection<String> authoritiesForClosedCpc = getActiveAuthorities(cpcClosed);
+      for (Map<String, ResourceSubscriber<? extends ResourceUpdate>> subscriberMap :
+          resourceSubscribers.values()) {
+        for (ResourceSubscriber<? extends ResourceUpdate> subscriber : subscriberMap.values()) {
+          if (subscriber.hasResult() || !authoritiesForClosedCpc.contains(subscriber.authority)) {
+            continue;
+          }
+
+          // try to fallback to lower priority control plane client
+          if (shouldTryFallback && manageControlPlaneClient(subscriber).didFallback) {
```

Comment:
Oh, I had meant to think more about the "I also suggest removing didFallback." thing before sending it out. Right now it is probably better to keep `didFallback`, because it is better to call subscriber.onError from the wrong cpc than never because the cpc was already failing.

Reply:
Since tracing through whether or not this was safe made MY brain hurt (and I'm not sick), I'm happy to keep it the way it is.

Reply:
> The only possible error is an unknown authority, which is already handled by Subscriber.addWatcher().

I'm realizing now that isn't the case. I see the old code had a try-catch around `getOrCreateControlPlaneClient(ServerInfo)` for `IllegalArgumentException`. I had thought that was a new catch for the new throw in `getOrCreateControlPlaneClient(String)`. The old line shows as covered, so some test is luckily exercising it. Seems it is handling the case where gRPC might throw when creating a transport.

Reply:
In `getOrCreateControlPlaneClient(String)` we need to catch `IllegalArgumentException` from `getOrCreateControlPlaneClient(ServerInfo)`, because we wouldn't want failure to create one transport to prevent us from trying other servers.

But handling that error is tricky. If we log and keep going then `addCpcToAuthority()` will add `null` to the activateCpClients authority list. We'd need to put protections everywhere it is used. I think we'd be better off making FailingXdsTransport (I know, I know. Hide your surprise). The main issue I see is we probably shouldn't call `eventHandler` until `StreamingCall.start()` returns, which requires a separate thread. If we move the IllegalStateException handling into GrpcXdsTransport then we could use `GrpcUtil.SHARED_CHANNEL_EXECUTOR`.

Ideas?